keps/sig-scheduling/4247-queueinghint/README.md

```diff
@@ -0,0 +1,1122 @@
+//   - `newObj` is nil if the event is delete event.
+type QueueingHintFn func(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (QueueingHint, error)
+
+type QueueingHint int
+
+const (
+	// QueueSkip implies that the cluster event has no impact on
+	// scheduling of the pod.
+	QueueSkip QueueingHint = iota
+
+	// Queue implies that the Pod may be schedulable by the event.
+	Queue
+)
+```
+
+### When to skip/not skip backoff
+
+BackoffQ is a light way of keeping throughput high 
+by preventing pods that are "permanently unschedulable" from blocking the queue.
+
+And, the more the Pod has been rejected in the scheduling cycle, the longer the Pod needs to wait as backoff.
+**We can regard the backoff as a penalty of wasting the scheduling cycle.**
+
+So, when, for example, NodeAffinity rejected the Pod and later returns `Queue` in its `QueueingHintFn`,
+the Pod's scheduling is retried after going through the backoff. 
+It's because the past scheduling cycle was wasted by that Pod.
+
+But, some plugins need to go through some failures in the scheduling cycle by design.
+[DRA](https://github.com/kubernetes/kubernetes/tree/v1.27.6/pkg/scheduler/framework/plugins/dynamicresources) plugin is one example in in-tree plugins - at the Reserve extension point, it tells the resource driver the scheduling result, and rejects the Pod once to wait for the response from the resource driver.
+In this kind of rejections, we cannot say the scheduling cycle is wasted because the scheduling result from it is used to proceed the Pod's scheduling forward, that particular scheduling cycle is failed though.
+So, Pods rejected by such reasons don't need to suffer a penalty (backoff).
+
+In order to support such cases, we introduces a new status `Pending`.
+When the `DRA` plugin rejected the Pod with `Pending` and later returns `Queue` in its `QueueingHintFn`,
+the pod skips the backoff and the Pod's scheduling is retried. 
+
+### How QueueingHint is executed in the scheduling queue
+
+When the cluster event happens, the scheduling queue executes `QueueingHintFn`
+of plugins which rejected the Pod in a previous scheduling cycle.
+
+Here are some scenarios to describe how they're executed and how the Pod is moved.
+
+#### Pod rejected by one or more plugins
+
+Let's say there are three Nodes. 
+When the Pod goes to the scheduling cycle, one Node is rejected due to no enouch capacity, 
+other two Nodes are rejected because they don't match Pod's NodeAffinity.
+
+In this case, the Pod gets `NodeResourceFit` and `NodeAffinity` as unschedulable plugins,
+and it's put back to the unschedulable pod pool.
+
+After then, every time the cluster events registered in those plugins happen,
+the scheduling queue notifies them through QueueingHint.
+If either of QueueingHintFn from `NodeResourceFit` or `NodeAffinity` returns `Queue`,
+the Pod is moved to activeQ/backoffQ.
+(For example, when `NodeAdded` event happens, the QueueingHint of `NodeResourceFit` return `Queue`
```

Comment:
I'm not saying it's impossible to improve how hints and unsched plugins are handled in the scheduling queue like your idea in the future. Actually, I remember I mentioned a similar idea somewhere before, like:
1. When the Pod is rejected, the scheduling queue tracks which Node is rejected by which plugins. (map[Nodename][]PluginName)
2. When the scheduling queue receives the event, the scheduling queue executes QueueingHint. (same as now)
3. QueueingHint somehow returns which Node might be schedulable by this change. (like return map[NodeName]QueueingHint ?)
4. If one Node seems solve all rejection from plugin, requeue this Pod

Bear with me, it's a super rough idea. I know, for example, (3) is not simple to do with cross-Node constraints like PodTopologySpread.
But, anyway, that kind of enhancement could be a big enough change to be treated as another enhancement after the basis of QueueingHint, discussed in this KEP, is done.

Reply:
In the current implementation, you also can't know whether this will make podA schedulable, because podB label update may lead to podA(unschedulable B/C podTopologySpread plugin before) fit for any node in the cluster, unless you process the whole prefilter & filter, so for cross-node plugins, we may simply return `Queue` like we do today.

I think what I raised here is also not a mature idea, processing all the plugin hintFuncs is not necessary sometimes, like nodeA is only unschedulable for pluginA, no need to go through pluginB's hintFunc. As you mentioned, we may need more work for fine-gained control.  Just bring this out for collision of thoughts. Do anyone has other concerns on this?

Reply:
> for cross-node plugins, we may simply return Queue like we do today.

That means cross-node plugins would have to return `Queue` to any events they register. In the above my example, we need to make PodTopologySpread return Queue at (5) although that PodUpdated event has completely nothing to do with PodTopologySpread.  
So, always just returning `Queue` in such difficult occasions cannot be the answer to this.

And, No, we don't do this today because we focus on the transition.

> In the current implementation, you also can't know whether this will make podA schedulable

So, to be accurate, in the current design, we _don't need to know_ whether the event makes PodA schedulable "on a certain Node(s)". The current design is kind of loose on it. It focuses on the transition only, and doesn't check further. 
For example, in the current design, PodTopologySpread returns `Queue` when the assigned Pod gets changed like from unmatch to match with the PodTopologySpread's selector. We don't need to check further about whether this Pod's change makes some Nodes schedulable actually or not.  And the scheduling queue requires that Pod into active/backoffQ with `Queue` from PodTopologySpread.

On the other hand, in your design, we'd need to know how that event affects the possibility of scheduling the Pod into some Nodes actually. (otherwise we miss some possibility as the above my example shows) And, returning `Queue` only when we're sure a certain Node(s) can be schedulable after the event from all unsched plugins PoV, which is difficult to achieve in cross-node constraints. (unless, as you said, we process the whole prefilter / filter)

So, basically, I don't think we need to go away from the current design, which is focusing on the transition of the event.
(Also, I [_guess_](https://github.com/kubernetes/enhancements/pull/4256#discussion_r1349809958) we can achieve the similar stuff on the top of the current design.)

Reply:
> we need to make PodTopologySpread return Queue at (5) although that PodUpdated event has completely nothing to do with PodTopologySpread

To clarify, they do related because podTopologySpread cares about the number of pods with specified labels. What I mean *always return queue* still cares about the related events, but we can't know exactly the result until we run prefilter & filter.

Reply:
> What I mean always return queue still cares about the related events

We'd _**have to**_ return Queue to all non-Node events they register. Note that there could be also custom plugins which have Pod related events. Let's say the scheduler has one (weird) custom plugin named `WeirdPodAffinity` 😅  which does PodAffinity but with Pod's annotation instead of labels. 

Here's the scenario same as [this comment](https://github.com/kubernetes/enhancements/pull/4256#discussion_r1350289036), but using WeirdPodAffinity instead of PodAffinity.

1. In PodA's scheduling: NodeA is rejected by PodTopologySpread. NodeB is rejected by WeirdPodAffinity while it gets OK from PodTopologySpread.
2. PodA is queued back into the unsched pod pool with {unschedPlugins: PodTopologySpread / WeirdPodAffinity}.
3. PodB, which is already scheduled on NodeB, gets an update on its annotation. (AssignedPodUpdated)
4. WeirdPodAffinity returns Queue because PodB's new annotation matches with PodA's WeirdPodAffinity.
5. But, PodTopologySpread returns QueueSkip because PodB's update has nothing to do with PodA's PodTopologySpread.

PodTopologySpread needs to return `Queue` at (5) because this PodUpdated event makes NodeB schedulable.
So, from PodTopologySpread's PoV, it cannot know how other plugins could return Queue to Pod related events. And it needs to always return `Queue` when they receive PodUpdated.

Reply:
Yes, cross-nodes is complex ..